flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/Fabric8FlinkKubeClient.java

```diff
@@ -355,4 +372,33 @@ private int getServiceNodePort(Service service, ConfigOption<Integer> configPort
 		}
 		return port;
 	}
+
+	/**
+	 * The help class to get the executor to run asynchronous operations on. It could create a dedicated thread pool or
+	 * reuse an existing one.
+	 */
+	private static class ExecutorWrapper {
+
+		private final ExecutorService internalExecutorService;
+		private Executor executor = null;
+
+		ExecutorWrapper(Configuration flinkConfig) {
+			internalExecutorService = Executors.newFixedThreadPool(
+				flinkConfig.getInteger(KubernetesConfigOptions.CLIENT_ASYNC_THREAD_POOL_SIZE),
+				new ExecutorThreadFactory("FlinkKubeClient-IO"));
+		}
+
+		ExecutorWrapper(Configuration flinkConfig, Executor executor) {
+			this(flinkConfig);
+			this.executor = executor;
+		}
+
+		Executor getExecutor() {
+			return executor == null ? this.internalExecutorService : executor;
+		}
+
+		public void close() {
+			ExecutorUtils.gracefulShutdown(5, TimeUnit.SECONDS, this.internalExecutorService);
+		}
+	}
```

Comment:
An alternative is that we always have two final field

```
private final Executor execute;
private final Runnable onClose;
```

and compose the close manner where

```
() -> {} // for passed Executor
() -> ExecutorUtils.gracefulShutdown(5, TimeUnit.SECONDS, this.internalExecutorService); // for internal service
```

since we initialize `onClose` in constructor we know whether `executor` is a passed `Executor` or an internal initialized service.